### PR TITLE
LPD-35169 Bind published object definitions with entries

### PR DESCRIPTION
**Pull request by devpedrotavares**
Related tickets: [LPD-35169](https://liferay.atlassian.net/browse/LPD-35169), [LPD-36653](https://liferay.atlassian.net/browse/LPD-36653)

[LPD-35169]: https://liferay.atlassian.net/browse/LPD-35169?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[LPD-36653]: https://liferay.atlassian.net/browse/LPD-36653?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by marcelabc** [comment]
Just started reviewing :)

File: modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectRelationshipLocalServiceImpl.java [hunk 1]
```diff
@@ -1531,38 +1553,29 @@ private void _validateDeletionType(
...
 
 		if (!edge || !FeatureFlagManagerUtil.isEnabled("LPS-187142")) {
 			return;
 		}
 
-		if (!Objects.equals(
-				objectRelationship.getType(),
-				ObjectRelationshipConstants.TYPE_ONE_TO_MANY)) {
-
+		if (!type.equals(ObjectRelationshipConstants.TYPE_ONE_TO_MANY)) {
```
Comment (marcelabc) [inline modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectRelationshipLocalServiceImpl.java:1565]: Are you sure that `type` will not be Null at this point? Can this be null because the user did not send this information in the request?
Why did you change from` Objects.equals()`?

File: modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/service/test/ObjectRelationshipLocalServiceTest.java [hunk 2]
```diff
@@ -876,23 +876,94 @@ public void testBindPublishedObjectDefinitions() throws Exception {
 
 		// Bind two published object definitions
 
+		ObjectDefinition objectDefinitionB =
+			_addAndPublishCustomObjectDefinition("B");
+		ObjectDefinition objectDefinitionBB =
+			_addAndPublishCustomObjectDefinition("BB");
```
Comment (marcelabc) [inline modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/service/test/ObjectRelationshipLocalServiceTest.java:882]: Why did you change from A and AA?

File: modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/service/test/ObjectRelationshipLocalServiceTest.java [hunk 3]
```diff
@@ -876,23 +876,94 @@ public void testBindPublishedObjectDefinitions() throws Exception {
...
+			objectDefinitionBBB.getObjectDefinitionId());
+
+		AssertUtils.assertFailure(
+			ObjectRelationshipEdgeException.class,
+			"Object definitions in a root context must not form a cycle",
+			() -> _bindObjectDefinitions(
+				objectDefinitionBBB.getObjectDefinitionId(),
+				objectDefinitionB.getObjectDefinitionId()));
+
+		// Must not have more than one parent
```
Comment (marcelabc) [inline modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/service/test/ObjectRelationshipLocalServiceTest.java:935]: // Unable to bind the object definitions when the child object definition is bound to another object definition

**Comment by marcelabc** [comment]
The changes from Pedro was reverted :disappointed:

File: modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectRelationshipLocalServiceImpl.java [hunk 4]
```diff
@@ -1571,6 +1571,14 @@ private void _validateEdge(
 					"object definitions to be an edge of a root context");
 		}
 
+		if (!StringUtil.equals(
+				objectDefinition1.getScope(), objectDefinition2.getScope())) {
+
+			throw new ObjectRelationshipEdgeException(
+				"Object definitions in a root context must have the same " +
+					"scope");
+		}
```
[the remark this answers is not in the document]
Comment (devpedrotavares) [reply modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectRelationshipLocalServiceImpl.java]: Are you okay with "Unable to bind object definitions when they have different scopes"?

File: modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectRelationshipLocalServiceImpl.java [hunk 5]
```diff
@@ -1579,6 +1579,14 @@ private void _validateEdge(
 					"scope");
 		}
 
+		if ((objectDefinition1.getRootObjectDefinitionId() != 0) &&
+			(objectDefinition1.getRootObjectDefinitionId() ==
+				objectDefinition2.getObjectDefinitionId())) {
+
+			throw new ObjectRelationshipEdgeException(
+				"Object definitions in a root context must not form a cycle");
+		}
```
[the remark this answers is not in the document]
Comment (devpedrotavares) [reply modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectRelationshipLocalServiceImpl.java]: I changed it to the second one

File: modules/apps/notification/notification-test/src/testIntegration/java/com/liferay/notification/internal/type/test/EmailNotificationTypeTest.java [hunk 6]
```diff
@@ -1146,7 +1146,7 @@ private ObjectDefinition _addAndPublishCustomObjectDefinition(
 						null, TestPropsValues.getUserId(),
 						accountEntryObjectDefinition.getObjectDefinitionId(),
 						objectDefinition.getObjectDefinitionId(), 0,
-						ObjectRelationshipConstants.DELETION_TYPE_PREVENT,
+						ObjectRelationshipConstants.DELETION_TYPE_PREVENT, true,
```
[the remark this answers is not in the document]
Comment (devpedrotavares) [reply modules/apps/notification/notification-test/src/testIntegration/java/com/liferay/notification/internal/type/test/EmailNotificationTypeTest.java]: No, that was a mistake. Thank you for pointing it out.

File: modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectRelationshipLocalServiceImpl.java (hunk 1 above)
Comment (devpedrotavares) [reply modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectRelationshipLocalServiceImpl.java]: I think it can be null, I switched it back to using `Objects.equals()`. Thank you

File: modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectRelationshipLocalServiceImpl.java [hunk 7]
```diff
@@ -1531,38 +1553,29 @@ private void _validateDeletionType(
...
+		if (!type.equals(ObjectRelationshipConstants.TYPE_ONE_TO_MANY)) {
 			throw new ObjectRelationshipEdgeException(
 				"Object relationship must be one to many to be an edge of a " +
 					"root context");
 		}
 
-		if (objectRelationship.isSelf()) {
+		if (objectDefinition1.getObjectDefinitionId() ==
+				objectDefinition2.getObjectDefinitionId()) {
+
```
[the remark this answers is not in the document]
Comment (devpedrotavares) [reply modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectRelationshipLocalServiceImpl.java:1569]: Because I'm not passing the `objectRelationship` reference anymore. Since the method is also called on add.

File: modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/service/test/ObjectRelationshipLocalServiceTest.java (hunk 2 above)
Comment (devpedrotavares) [reply modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/service/test/ObjectRelationshipLocalServiceTest.java:882]: Because I am using them in `AssertUtils.assertFailure` which requires final variables. So I couldn't use A and AA directly.

**Comment by devpedrotavares** [comment]
> The changes from Pedro was reverted 😞 However, I'm unsure if it is worth rebasing with Master because maybe his modification will be small. Can you check with him please?

I talked to him and the changes are few. Perhaps it's best to wait for his PR to get merged

**Comment by marcelabc** [comment]
Just started reviewing :)

File: modules/apps/object/object-admin-rest-impl/src/main/java/com/liferay/object/admin/rest/internal/resource/v1_0/ObjectRelationshipResourceImpl.java [hunk 8]
```diff
@@ -211,6 +212,7 @@ public ObjectRelationship postObjectDefinitionObjectRelationship(
 				GetterUtil.getLong(
 					objectRelationship.getParameterObjectFieldId()),
 				objectRelationship.getDeletionTypeAsString(),
+				objectRelationship.getEdge(),
```
Comment (marcelabc) [inline modules/apps/object/object-admin-rest-impl/src/main/java/com/liferay/object/admin/rest/internal/resource/v1_0/ObjectRelationshipResourceImpl.java:215]: Can you use `GetterUtil.getBoolean(objectRelationship.getEdge())` ?

File: modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectEntryLocalServiceImpl.java [hunk 9]
```diff
@@ -1628,6 +1632,119 @@ public ObjectEntry updateObjectEntry(
...
+			ObjectDefinition objectDefinition =
+				_objectDefinitionPersistence.findByPrimaryKey(
+					node.getPrimaryKey());
+
+			Indexer<ObjectEntry> indexer = IndexerRegistryUtil.getIndexer(
+				objectDefinition.getClassName());
+
+			_performActionableDynamicQuery(
+				objectDefinition1.getObjectDefinitionId(), true,
+				(ObjectEntry objectEntry) -> indexer.reindex(objectEntry));
```
Comment (marcelabc) [inline modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectEntryLocalServiceImpl.java:1742]: _performActionableDynamicQuery(
	**objectDefinition1**.getObjectDefinitionId(), true,
	(ObjectEntry objectEntry) -> indexer.reindex(objectEntry));

Double-checking: is this supposed to always be the **objectDefinition1** even when the `iterator.next()` changes to another object definition of the objectDefinitionTree?

File: modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectRelationshipLocalServiceImpl.java [hunk 10]
```diff
@@ -1518,45 +1547,81 @@ private void _validateDeletionType(
...
+						"relationship can be an edge to a root context"));
+			}
+		}
+
+		if (!StringUtil.equals(
+				objectDefinition1.getScope(), objectDefinition2.getScope())) {
+
+			throw new ObjectRelationshipEdgeException(
+				"Unable to bind object definitions when they have different " +
+					"scopes");
```
Comment (marcelabc) [inline modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectRelationshipLocalServiceImpl.java:1613]: "Unable to bind **the** object definitions when they have different "

**Review by marcelabc — CHANGES_REQUESTED** [review]
Hi 
I finished my review. Let me know if you have questions :smile:

File: modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/service/test/ObjectRelationshipLocalServiceTest.java [hunk 11]
```diff
@@ -872,23 +876,234 @@ public void testBindPublishedObjectDefinitions() throws Exception {
...
 					).put(
-						"AA", new String[0]
+						"BB", new String[0]
 					).build(),
 					_treeFactory.createObjectDefinitionTree(
 						objectDefinition1.getObjectDefinitionId(),
 						_objectDefinitionLocalService::getObjectDefinition),
 					_objectDefinitionLocalService));
 
+		// The object definitions must have the same scope
```
Comment (marcelabc) [inline modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/service/test/ObjectRelationshipLocalServiceTest.java:898]: // Unable to bind object definitions when they have different scopes"

I think it is best to follow the exception message because it explains better;

The actual message needs to clarify what we are trying to do so that the object definitions must have the same scope 
// The object definitions must have the same scope
> for what?

File: modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/service/test/ObjectRelationshipLocalServiceTest.java [hunk 12]
```diff
@@ -872,23 +876,234 @@ public void testBindPublishedObjectDefinitions() throws Exception {
...
+
+		AssertUtils.assertFailure(
+			ObjectRelationshipEdgeException.class,
+			"Unable to bind object definitions when they have different scopes",
+			() -> _bindObjectDefinitions(
+				objectDefinitionB.getObjectDefinitionId(),
+				siteObjectDefinition.getObjectDefinitionId()));
+
+		// The object relationship must not create a circular reference in
+		// a root context
```
Comment (marcelabc) [inline modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/service/test/ObjectRelationshipLocalServiceTest.java:920]: Following the same principle, what do you think about renaming to:
// Unable to bind the object definitions because the object relationship must not create a circular reference in a root context

File: modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectEntryLocalServiceImpl.java (hunk 9 above)
Comment (devpedrotavares) [reply modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectEntryLocalServiceImpl.java]: That was inattention, I forgot to update it from a previous logic. 

Also, while analyzing this part again, I noticed we do not always have to reindex `objectDefinition1`, so I added a conditional to avoid doing so unnecessarily on this [commit](https://github.com/liferay-objects/liferay-portal/pull/3682/commits/bd10fb0fb4913ffa716e70c73ae7f9210ef97c51).

**Review by marcelabc — CHANGES_REQUESTED** [review]
Hi 
I finished my review. Let me know if you have questions :smile:
I Slack you with other investigations to do! Tomorrow we can work together :+1:

**Comment by devpedrotavares** [comment]
Hi @marcelabc , 
I rebased on top of https://github.com/liferay-objects/liferay-portal/pull/3704

**Comment by devpedrotavares** [comment]
ci:test:sf

**Comment by devpedrotavares** [comment]
Hi @marcelabc , I rebased on master, since Pedro's commits were merged

**Comment by marcelabc** [comment]
Hi @devpedrotavares,
Thank you for all the effort you put into this PR. 
Since the team is trying to forward the PRs to Brian quickly, I created another task to review and evolve your solution for the Tree validation to check its height.
I asked you to include it but did not manage to review it on time. It is my fault. So, I reviewed the code from our last review and resent another PR. Please check it [here](https://github.com/liferay-objects/liferay-portal/pull/3711).

**Comment by marcelabc** [comment]
The new [subtask](https://liferay.atlassian.net/browse/LPD-38680). I will align with you on who should be responsible for moving forward with this validation.
